server/src/main/java/org/elasticsearch/cluster/coordination/LagDetector.java

```diff
@@ -0,0 +1,156 @@
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.cluster.coordination;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.threadpool.ThreadPool.Names;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+
+import static org.elasticsearch.common.util.concurrent.ConcurrentCollections.newConcurrentMap;
+
+/**
+ * A publication can succeed and complete before all nodes have applied the published state and acknowledged it; however we need every node
+ * eventually either to apply the published state (or a later state) or be removed from the cluster. This component achieves this by
+ * removing any lagging nodes from the cluster after a timeout.
+ */
+public class LagDetector {
+
+    private static final Logger logger = LogManager.getLogger(LagDetector.class);
+
+    // the timeout for each node to apply a value after the end of publication
+    public static final Setting<TimeValue> CLUSTER_STATE_APPLICATION_TIMEOUT_SETTING =
+        Setting.timeSetting("cluster.applier.timeout",
+            TimeValue.timeValueMillis(60000), TimeValue.timeValueMillis(1), Setting.Property.NodeScope);
+
+    private final TimeValue clusterStateApplicationTimeout;
+    private final Consumer<DiscoveryNode> onLagDetected;
+    private final ThreadPool threadPool;
+    private final Map<DiscoveryNode, NodeAppliedStateTracker> appliedStateTrackersByNode = newConcurrentMap();
+
+    public LagDetector(final Settings settings, final ThreadPool threadPool, final Consumer<DiscoveryNode> onLagDetected) {
+        this.threadPool = threadPool;
+        this.clusterStateApplicationTimeout = CLUSTER_STATE_APPLICATION_TIMEOUT_SETTING.get(settings);
+        this.onLagDetected = onLagDetected;
+    }
+
+    public void setTrackedNodes(final Iterable<DiscoveryNode> discoveryNodes) {
+        final Set<DiscoveryNode> discoveryNodeSet = new HashSet<>();
+        discoveryNodes.forEach(discoveryNodeSet::add);
```

Comment:
let's ignore the local node. This also makes writing the assertions easier in Coordinator.invariant()